rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala

```diff
@@ -205,29 +205,29 @@ class DebruijnInterpreter[M[_], F[_]](
 
   private def eval(
       term: GeneratedMessage
-  )(
-      implicit env: Env[Par],
-      rand: Blake2b512Random
-  ): M[Unit] =
-    term match {
-      case term: Send    => reportErrors(eval(term))
-      case term: Receive => reportErrors(eval(term))
-      case term: New     => reportErrors(eval(term))
-      case term: Match   => reportErrors(eval(term))
-      case term: Bundle  => reportErrors(eval(term))
-      case term: Expr =>
-        term.exprInstance match {
-          case e: EVarBody    => reportErrors(eval(e.value.v) >>= (eval(_)))
-          case e: EMethodBody => reportErrors(evalExprToPar(e) >>= (eval(_)))
-          case other          => BugFoundError(s"Undefined term: \n $other").raiseError[M, Unit]
-        }
-      case other => BugFoundError(s"Undefined term: \n $other").raiseError[M, Unit]
+  )(implicit env: Env[Par], rand: Blake2b512Random): M[Unit] =
+    reportErrors {
+      term match {
+        case term: Send    => eval(term)
+        case term: Receive => eval(term)
+        case term: New     => eval(term)
+        case term: Match   => eval(term)
+        case term: Bundle  => eval(term)
+        case term: Expr =>
+          term.exprInstance match {
+            case e: EVarBody    => eval(e.value.v) >>= (eval(_))
+            case e: EMethodBody => evalExprToPar(e) >>= (eval(_))
+            case other          => BugFoundError(s"Undefined term: \n $other").raiseError[M, Unit]
+          }
+        case other => BugFoundError(s"Undefined term: \n $other").raiseError[M, Unit]
+      }
     }
 
   private def reportErrors(process: M[Unit]): M[Unit] =
     process.handleErrorWith {
-      case e @ OutOfPhlogistonsError => e.raiseError[M, Unit]
-      case e                         => fTell.tell(e)
+      case error @ OutOfPhlogistonsError => error.raiseError[M, Unit]
+      case error: InterpreterError       => fTell.tell(error)
+      case error                         => error.raiseError[M, Unit]
```

Comment:
Why don't we just `.raiseError` in all cases? Did we try that approach in the 'error handling vs tuplespace corruption' investigation?

Reply:
My best guess (and it must be a guess because there's no documentation) is that `Vector[InterpreterError]` is used to report errors so that errors can be reliably compared upon replay, despite the fact that errors are not compared upon replay 🥇. We can probably remove the error-log and just use the `raiseError` channel, but I think that should be delayed until after deploy charging is in.